### PR TITLE
Filter out captured primary constructor parameters from completion lists.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/69300.

Core issue is that we do not want to show the users these members as it will cause an easy to miss warning that the parameter they're capturing (either by passing to the base, or assigning to a member) will then allocate another field to store the capture as well.

The user should instead use the real field/property the user has declared instead.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@genlu  ptal.

File: src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs [hunk 1]
```diff
@@ -297,27 +298,124 @@ private ImmutableArray<ISymbol> GetSymbolsForExpressionOrStatementContext()
                 ? _context.SemanticModel.LookupStaticMembers(_context.LeftToken.SpanStart)
                 : _context.SemanticModel.LookupSymbols(_context.LeftToken.SpanStart);
 
-            // filter our top level locals if we're inside a type declaration.
-            if (_context.ContainingTypeDeclaration != null)
-                symbols = symbols.WhereAsArray(s => s.ContainingSymbol.Name != WellKnownMemberNames.TopLevelStatementsEntryPointMethodName);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs:302]: we were realizing a filtered list over and over again.  Tweaked the filtering to be single pass.

File: src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs [hunk 2]
```diff
@@ -297,27 +298,124 @@ private ImmutableArray<ISymbol> GetSymbolsForExpressionOrStatementContext()
...
+                CSharpSyntaxContext context,
+                INamedTypeSymbol? enclosingNamedType,
+                ISet<INamedTypeSymbol> outerTypes,
+                bool filterOutOfScopeLocals,
+                ISymbol symbol,
+                CancellationToken cancellationToken)
+            {
+                // filter our top level locals if we're inside a type declaration.
+                if (context.ContainingTypeDeclaration != null && symbol.ContainingSymbol.Name == WellKnownMemberNames.TopLevelStatementsEntryPointMethodName)
+                    return true;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs:320]: existing check.  just moved into the single filter method.

File: src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs [hunk 3]
```diff
@@ -297,27 +298,124 @@ private ImmutableArray<ISymbol> GetSymbolsForExpressionOrStatementContext()
...
-            // should not be included in the completion list, so remove those. Filter them away,
-            // unless we're in the debugger, where we show all locals in scope.
-            if (filterOutOfScopeLocals)
-                symbols = symbols.WhereAsArray(symbol => !symbol.IsInaccessibleLocal(_context.Position));
+                if (symbol.IsExtensionMethod() &&
+                    !Equals(enclosingNamedType, symbol.ContainingType) &&
+                    !outerTypes.Any(outerType => outerType.Equals(symbol.ContainingType)))
+                {
+                    return true;
+                }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs:327]: existing check.  just moved into the single filter method.

File: src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs [hunk 4]
```diff
@@ -297,27 +298,124 @@ private ImmutableArray<ISymbol> GetSymbolsForExpressionOrStatementContext()
...
 
-            return symbols;
+                // The symbols may include local variables that are declared later in the method and should not be
+                // included in the completion list, so remove those. Filter them away, unless we're in the debugger,
+                // where we show all locals in scope.
+                if (filterOutOfScopeLocals && symbol.IsInaccessibleLocal(context.Position))
+                    return true;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs:333]: existing check.  just moved into the single filter method.

**Comment by CyrusNajmabadi** [comment]
@genlu ptal.

File: src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs [hunk 5]
```diff
@@ -297,27 +298,124 @@ private ImmutableArray<ISymbol> GetSymbolsForExpressionOrStatementContext()
...
+                    return false;
+                }
+
+                // We're not in an initializer.  Filter out this primary constructor parameter if it's already been
+                // captured by an existing field or property initializer.
+
+                var parameterName = parameterSymbol.Name;
+                foreach (var reference in enclosingNamedType.DeclaringSyntaxReferences)
+                {
+                    if (reference.GetSyntax(cancellationToken) is not TypeDeclarationSyntax typeDeclaration)
```
Comment (genlu) [inline src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs:373]: When would this be true?

**Review by genlu — APPROVED** [review]
:shipit:

File: src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs:373]: Should be true. But I like being cautious :-)

**Pull request merged**